server.go

```diff
@@ -61,36 +75,180 @@ func Serve(ctx context.Context, c *websocket.Conn, execer Execer) error {
+					if err != nil {
+						cancel()
+						return err
+					}
+
+					// Default to buffer 64KB.
+					ringBuffer, err := circbuf.NewBuffer(64 * 1024)
+					if err != nil {
+						cancel()
+						return xerrors.Errorf("unable to create ring buffer %w", err)
+					}
+
+					rprocess = &reconnectingProcess{
+						activeConns: make(map[string]net.Conn),
+						process:     process,
+						// Timeouts created with AfterFunc can be reset.
+						timeout:    time.AfterFunc(options.ReconnectingProcessTimeout, cancel),
+						ringBuffer: ringBuffer,
+					}
+					reconnectingProcesses.Store(header.ID, rprocess)
+
+					// If the process exits send the exit code to all listening
+					// connections then close everything.
+					go func() {
+						err = process.Wait()
+						code := 0
+						if exitErr, ok := err.(ExitError); ok {
+							code = exitErr.Code
+						}
+						rprocess.activeConnsMutex.Lock()
+						for _, conn := range rprocess.activeConns {
+							_ = sendExitCode(ctx, code, conn)
+						}
+						rprocess.activeConnsMutex.Unlock()
+						rprocess.Close()
+						reconnectingProcesses.Delete(header.ID)
+					}()
+
+					// Write to the ring buffer and all connections as we receive stdout.
+					go func() {
+						buffer := make([]byte, 32*1024)
+						for {
+							read, err := rprocess.process.Stdout().Read(buffer)
+							if err != nil {
+								// When the process is closed this is triggered.
+								break
+							}
+							part := buffer[:read]
+							_, err = rprocess.ringBuffer.Write(part)
+							if err != nil {
+								flog.Error("reconnecting process %s write buffer: %v", header.ID, err)
+								cancel()
+								break
+							}
+							rprocess.activeConnsMutex.Lock()
+							for _, conn := range rprocess.activeConns {
+								_ = sendOutput(ctx, part, conn)
+							}
+							rprocess.activeConnsMutex.Unlock()
+						}
+					}()
 				}
-				_ = sendExitCode(ctx, 0, wsNetConn)
-			}()
+
+				err = sendPID(ctx, process.Pid(), wsNetConn)
+				if err != nil {
+					flog.Error("failed to send pid %d", process.Pid())
+				}
+
+				// Write out the initial contents in the ring buffer.
+				err = sendOutput(ctx, rprocess.ringBuffer.Bytes(), wsNetConn)
+				if err != nil {
+					return xerrors.Errorf("write reconnecting process %s buffer: %w", header.ID, err)
+				}
+
+				// Store this connection on the reconnecting process.  All connections
+				// stored on the process will receive the process's stdout.
+				connectionID := uuid.NewString()
+				rprocess.activeConnsMutex.Lock()
+				rprocess.activeConns[connectionID] = wsNetConn
+				rprocess.activeConnsMutex.Unlock()
+
+				// Keep resetting the inactivity timer while this connection is alive.
+				rprocess.timeout.Reset(options.ReconnectingProcessTimeout)
+				heartbeat := time.NewTimer(options.ReconnectingProcessTimeout / 2)
+				defer heartbeat.Stop()
+				go func() {
+					for {
+						select {
+						case <-heartbeat.C:
+						}
```

Comment:
This should create a context that gets canceled on defer, otherwise this goroutine will leak forever.

eg. `heartbeat.Stop()` gets called, so the channel returns instantly. The for loop will run forever!